src/unit.c

```diff
@@ -23,148 +23,44 @@
 #include <float.h>
 #include <string.h>
 
-int isUnitArithmetic(SEXP ua) {
-    return inherits(ua, "unit.arithmetic");
-}
-
-int isUnitList(SEXP ul) {
-    return inherits(ul, "unit.list");
-}
-
 /* Function to build a single-value unit SEXP internally.
  * Cannot build units requiring data as yet.
  */
 SEXP unit(double value, int unit) 
 {
-    SEXP u, units, classname;
-    PROTECT(u = ScalarReal(value));
-    PROTECT(units = ScalarInteger(unit));
-    /* NOTE that we do not set the "unit" attribute */
-    setAttrib(u, install("valid.unit"), units);
-    setAttrib(u, install("data"), R_NilValue);
-    PROTECT(classname = mkString("unit"));
-    classgets(u, classname);
-    UNPROTECT(3);
-    return u;
+  SEXP units = PROTECT(allocVector(VECSXP, 1));
+  SEXP u = SET_VECTOR_ELT(units, 0, allocVector(VECSXP, 3));
+  SET_VECTOR_ELT(u, 0, ScalarReal(value));
+  SET_VECTOR_ELT(u, 1, R_NilValue);
+  SET_VECTOR_ELT(u, 2, ScalarInteger(unit));
+  classgets(units, mkString("unit"));
+  UNPROTECT(1);
+  return units;
 }
 
 /* Accessor functions for unit objects
  */
 
 /* 
  * This is an attempt to extract a single numeric value from
- * a unit.  This is ONLY designed for use on "simple" units
- * (i.e., NOT unitLists or unitArithmetics)
+ * a unit.
  */
 double unitValue(SEXP unit, int index) {
-    /* Recycle values if necessary (used in unit arithmetic)
-     */
-    int n = LENGTH(unit);
-    return numeric(unit, index % n);
+  return Rf_asReal(VECTOR_ELT(VECTOR_ELT(unit, index % unitLength(unit)), 0));
```

Comment:
It feels like these would be easier to read if you extracted `VECTOR_ELT(unit, index % unitLength(unit)` out.

Are you sure the coercions (`Rf_asReal()`) etc are necessary? Couldn't it just be `REAL()[0]`?

Reply:
I thought `Rf_asReal` was made for these cases where you want to convert a single-length vector to a scalar?

Reply:
Wow - had not seen how much Rf_as*() did in addition to extracting the first element